### PR TITLE
feat: module unit tests can allow calls within module

### PR DESCRIPTION
**Pull request by matt2e**
We already have a way of providing verb mocks/fakes with `ftltest.WhenVerb(...)` but sometimes with tests it can be useful to allow calls to work fine within the module.
This PR enables this behaviour though it has to be explicitly enabled by the test. This allows the person writing the test to better understand what their unit test is touching.

---
Below is a rundown of how ftltest works for a test writer in practice:
_⌨️ Write module `echo` which has verb `Echo` and `InnerEcho`. `Echo` calls `InnerEcho`, `InnerEcho` calls `time.Time`_

_⌨️ Write the following test:_
```
func TestEcho(t *testing.T) {
  ctx := ftltest.Context(
    ftltest.WithConfig(defaultName, "fallback"),
  )
  resp, err := Echo(ctx, EchoRequest{Name: ftl.Some("world")})
  assert.NoError(t, err)
  assert.Equal(t, "Hello, world!!! It is 2024-01-01 00:00:00 +0000 UTC!", resp.Message)
}
```
🤖 Error:
> echo.innerEcho: no mock found: provide a mock with ftltest.WhenVerb(innerEcho, ...) or enable all calls within the module with ftltest.WithCallsAllowedWithinModule()

_⌨️ Add `ftltest.WithCallsAllowedWithinModule()` to `ftltest.Context()` options_
🤖 Error:
> echo.innerEcho: time.time: no mock found: provide a mock with ftltest.WhenVerb(time.time, ...)

_⌨️ Add the following to `ftltest.Context()` options_:
```
ftltest.WhenVerb(time.Time, func(ctx context.Context, req time.TimeRequest) (time.TimeResponse, error) {
  return time.TimeResponse{Time: stdtime.Date(2024, 1, 1, 0, 0, 0, 0, stdtime.UTC)}, nil
})
```
🤖 success

### CODE REVIEW

**Comment by alecthomas** [comment]
Niiiice, love the error messages.

**Review by alecthomas — APPROVED** [review]

**Review by alecthomas — COMMENTED** [review]
Add some integration tests for this.

File: go-runtime/ftl/ftltest/ftltest.go [hunk 1]
```diff
@@ -105,3 +105,21 @@ func WhenVerb[Req any, Resp any](verb ftl.Verb[Req, Resp], fake func(ctx context
...
+// Any overrides provided by calling WhenVerb(...) will take precedence
+func WithCallsAllowedWithinModule() func(context.Context) error {
+	return func(ctx context.Context) error {
+		overrider, ok := ftl.CallOverriderFromContext(ctx)
+		if !ok {
+			return fmt.Errorf("could not allow calls wihtin module, context not set up with call overrider")
+		}
+		mockProvider, ok := overrider.(*mockVerbProvider)
+		if !ok {
+			return fmt.Errorf("could not allow calls wihtin module, call overrider is not a MockProvider")
```
Comment (alecthomas) [inline go-runtime/ftl/ftltest/ftltest.go:120]:
```suggestion
			return fmt.Errorf("could not allow calls within module, call overrider is not a MockProvider")
```

File: go-runtime/ftl/call_overrider.go [hunk 2]
```diff
@@ -5,7 +5,7 @@ import (
 )
 
 type CallOverrider interface {
```
Comment (alecthomas) [inline go-runtime/ftl/call_overrider.go:7]: Can you document this interface and its purpose please, including what "inline" does, and what "override" does.

**Comment by matt2e** [comment]
Converting this to a draft for now. Could not get integration tests to work and it seems like removing the `configuration` package from go-runtime will greatly help.

Current status of this branch: new integration test is broken. Will revisit

**Comment by deniseli** [comment]
sorry for the review delay - approved! btw I really like how this all looks in internal/modulecontext/module_context.go after your more recent changes :)

**Pull request merged**